### PR TITLE
Add platform component

### PR DESCRIPTION
**Pull request by SergioEstevao**
## Description

This PR adds a platform component to allow us to make code adapt to mobile or web depending of the platform is running.

For the web it creates a component that always returns web and on iOS and Android it simple wraps around the RN component.

## How has this been tested?

There is still no uses for the component. But there is a couple of impending PRs that will use this functionality.

## Types of changes

New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/element/src/platform.js
Comment (youknowriad) [inline packages/element/src/platform.js]: What is this function for?

Comment (SergioEstevao) [reply packages/element/src/platform.js]: It's to write this kind of code you can see here: https://facebook.github.io/react-native/docs/platform-specific-code#docsNav

Comment (gziolo) [reply packages/element/src/platform.js]: This is one of the techniques used in the cross-platform projects to avoid extracting very small utilities to their own file just to override them depending on the platform. 

@SergioEstevao - can you find a place in code where it could be used so we had a real-life example.

File: packages/element/src/platform.js
Comment (koke) [inline packages/element/src/platform.js]: Since this seems like a copy of https://github.com/necolas/react-native-web/blob/0e81c6ef2758d4ca9b2099b1d04a4f8c417f0f43/packages/react-native-web/src/exports/Platform/index.js let's try to keep the original copyright attribution here.

File: packages/element/src/platform.js
Comment (koke) [reply packages/element/src/platform.js]: @mkevins was sharing an example from gallery this morning, where the placeholder label would be different since we don't support drag and drop:

```jsx
const placeholderLabel = Platform.select({
    native: __( 'Add media' ),
    web: __( 'Drag images, upload new ones or select files from your library.' ),
} );
```

Comment (koke) [reply packages/element/src/platform.js]: I spoke too soon, I think `Platform.select` [doesn't support](https://github.com/facebook/react-native/blob/master/Libraries/Utilities/Platform.ios.js#L62) the `native` specifier. It would be a nice addition to our implementation.

Comment (SergioEstevao) [reply packages/element/src/platform.js]: You want the native option to be added for the web component and the native component too, correct?

Comment (SergioEstevao) [reply packages/element/src/platform.js]: @koke something like this for native: 

```
import { Platform as OriginalPlatform} from 'react-native';

const Platform = {
	OS: 'native',
	select: ( obj ) => ( 'native' in obj ? obj.native : OriginalPlatform.select(obj) ),
};

export default Platform;```

Comment (mkevins) [reply packages/element/src/platform.js]: Nice :smiley: :+1:

Comment (koke) [reply packages/element/src/platform.js]: I think `Platform.OS` should always return the more specific value (`ios` or `android`). For select to work, I think it might be easier to just add our own implementation for both platforms (and/or submit a PR to react-native). The logic for iOS would be:

```js
const Platform = {
    OS: 'ios',
    select: (spec): => 'ios' in spec ? spec.ios : 'native' in spec ? spec.native : spec.default,
}
```

`Platform` exposes other things like OS version, so we should still inherit those from the react-native implementation.

We could also not implement `native` and just use `web`/`default` to differentiate, but I think that would break expectations since we're very used to having `.native.js` files now.

Comment (SergioEstevao) [reply packages/element/src/platform.js]: On my code above can I do this: 

`OriginalPlatform.select = (spec): => 'ios' in spec ? spec.ios : 'native' in spec ? spec.native : spec.default`

File: packages/element/src/platform.js
Comment (gziolo) [reply packages/element/src/platform.js]: 👍

File: packages/element/src/platform.js
Comment (gziolo) [inline packages/element/src/platform.js]: It would be great to include an example in the docs as well, since this implementation differs a bit, right? Regardless, it can change upstream, so we should document it.

Can we add simple unit tests that would also act like documentation? I personally, often look at unit tests when trying to learn some 3rd party software.

Comment (SergioEstevao) [reply packages/element/src/platform.js]: @gziolo just added both, give it a look to see if it's ok now :)

Comment (koke) [reply packages/element/src/platform.js]: BTW, I sent a PR upstream

https://github.com/facebook/react-native/pull/26966

I was going to suggest we could get the unit tests from there, but you added them faster 😅

Comment (koke) [reply packages/element/src/platform.js]: BTW, I sent a PR upstream

https://github.com/facebook/react-native/pull/26966

I was going to suggest we could get the unit tests from there, but you added them faster 😅

File: packages/element/README.md
Comment (koke) [inline packages/element/README.md]: I think it's fine to link to the react-native docs, but I'd prefer to document explicitly, since we only support `OS` and `select` on the web.

File: packages/element/src/platform.js
Comment (gziolo) [inline packages/element/src/platform.js]: How about:
```suggestion
 * This is the same concept as the React Native implementation.
  * @see https://facebook.github.io/react-native/docs/platform-specific-code#platform-module
```

**Comment by gziolo** [comment]
Nice work @SergioEstevao, let's sort out documentation and land it to see it in action. I hope it will simplify branching between platforms 👍

**Review by koke — APPROVED** [review]
👏 So happy to have this on Gutenberg 👏

File: packages/element/src/index.js
Comment (koke) [inline packages/element/src/index.js]: ESLint is complaining about this one

```suggestion
export { default as Platform } from './platform';
```

File: packages/element/src/platform.js
Comment (gziolo) [inline packages/element/src/platform.js]: One more nit before you merge:
```suggestion
import { Platform } from '@wordpress/element'; 

 * const placeholderLabel = Platform.select( {
```
to make it easier to copy for folks.

**Review by gziolo — APPROVED** [review]
:shipit:

**Comment by gziolo** [comment]
`npm run docs:build` is necessary to make Travis happy in addition to the JSDoc change :)

I hope we will automate it and leave it to the bot which does it automatically.

**Comment by mkevins** [comment]
Nice work @SergioEstevao ! :tada:

**Review by aduth — COMMENTED** [review]
One thing I might worry about with this specific implementation is that we're evaluating every possible value, knowing that ultimately only one will actually be used. Depending how it's used, this might not have much (or any) overhead, but I could consider it as optimized to allow unnecessary overhead all the same.

I wonder if instead (or even as a possible _optional_ enhancement), these could be specified as functions, so that only the function applicable to the current context is invoked.

```js
const result = Platform.select( {
    web: () => getComputationallyExpensiveValue(),
    // Or:
    // web: getComputationallyExpensiveValue,
    native: () => getOtherComputationallyExpensiveValue(),
} );
```

---

Having said that, and as I read the conversation history of the pull request before hitting "Submit", I see that this is based on some prior art from React Native. I think there's a compelling argument of consistency for consistency's sake, in case a developer already has some familiarity with this implementation.

**Comment by mkevins** [comment]
Hi @aduth :wave: 

> One thing I might worry about with this specific implementation is that we're evaluating every possible value, knowing that ultimately only one will actually be used.

That's a great point. My only concern with the proposed _optional_ enhancement would be that we may want to use `Platform.select` to select a _function_, so I suppose we'd need a mechanism to distinguish when we are using the interface in its original way, and when we want the callback style interface.

On a related note, the use of `Platform.OS` from `@wordpress/element` and from `react-native` has some inconsistencies arising from the fact that `Platform.OS` is a single primitive, and not suited to represent categorical hierarchies. E.g.:

Platform|@wordpress/element `Platform.OS ===`|react-native `Platform.OS ===`
-|-|-
Web|`'web'`|N/A
Android|`'native'`|`'android'`
iOS|`'native'`|`'ios'`

I.e. we "lose" some specificity when using the `Platform.OS` constant from @wordpress/element. This _could_ result in some confusion in the future if / when `*.js` and `*.native.js` files are being consolidated.

I believe `Platform.OS` is mostly used with `===` to branch behavior by platform (e.g. `if (Platform.OS === 'ios') { doSomethingSpecial(); }`), so I wonder if it'd be useful to include a new pattern of boolean flags as well: `if (Platform.isIos) { doSomethingSpecial(); }`. In this way, flags at different "levels" of the categorical hierarchy can coexist, e.g. `if (Platform.isNative) { doSomething(); }` would be possible with the same Platform import as `if (Platform.isIos) { doSomething(); }`

**Comment by koke** [comment]
> I wonder if instead (or even as a possible _optional_ enhancement), these could be specified as functions, so that only the function applicable to the current context is invoked.

@aduth That's a great point, but I believe there's nothing stopping you from doing that with the current implementation. I can't find an example right now, but I've seen some uses of the react-native one where functions were passed, and then they just added a `()` at the end of the `Platform.select` to return the value:

```js
const result = Platform.select( {
    web: () => getComputationallyExpensiveValue(),
    // Or:
    // web: getComputationallyExpensiveValue,
    native: () => getOtherComputationallyExpensiveValue(),
} )();
```

**Comment by aduth** [comment]
@mkevins @koke Thanks for the context! I see now that at this point, we can't really easily overload to add the function option since, as you mention, someone may choose to select a function result. It would have been something necessary to make the _only_ option, which has some small downsides of its own (_slightly_ less usable).

@mkevins I could imagine such constants could be nice too, since I imagine most of these use-cases could otherwise boil down to simple ternaries?

```js
const result = Platform.select( {
    web: 'foo',
    native: 'bar',
} );

// vs.

const result = Platform.isNative ? 'bar' : 'foo';
```

@koke Yeah, I could see how that might work. I was also imagining it might work to use [getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get):

```js
const result = Platform.select( {
    get web() {
        return getComputationallyExpensiveValue();
    },
    get native() {
        return getOtherComputationallyExpensiveValue();
    },
} );
```

My only problem with either of these is that because this isn't optimized as the primary workflow, it would be easily overlooked if someone is not mindful of the performance impact. Usually I would prefer if the framework would make those decisions on behalf of the developer (via the default interface).

**Comment by koke** [comment]
Every example I could find from react-native except one used `Platform.select` with simple values, most often to set slightly different styles per platform. The issue is possible but seems unlikely to me.

**Pull request merged**
